### PR TITLE
Created a starter script for fine tuning on whisper data. 

### PR DESCRIPTION
**Pull request by arjunpatel7**
I threw together a script to fine tune on @kpriyanshu256 's whisper transcripts using distilgpt2. The script is written such that we can easily pick a different model checkpoint, dataset, epochs, and block size. Could be useful for testing parts of our pipeline.

### CODE REVIEW

File: starter_training_script.ipynb [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+   "outputs": [],
+   "source": [
+    "import pandas as pd\n",
+    "import tensorflow as tf\n",
+    "from datasets import load_dataset\n",
+    "from transformers import AutoTokenizer, create_optimizer, AdamWeightDecay, TFAutoModelForCausalLM\n",
+    "\n",
+    "\n",
+    "#most code is taken from the old huggingface script for language modeling with \n",
+    "DATASET_PATH = \"kpriyanshu256/whisper-transcripts\"\n",
```
Comment (matallanas) [inline starter_training_script.ipynb]: Can you put the parameters on another cell just ofr example to maybe use it with papermill or other library to automate the execution?

File: starter_training_script.ipynb [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 62,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# create functions to read in JSON data of transcripts\n",
+    "\n",
+    "def retrieve_dataset(path):\n",
```
Comment (matallanas) [inline starter_training_script.ipynb]: No need of this function you only used once to load the dataset. You can do it directly on the cell below

File: starter_training_script.ipynb [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+    "    # given filepath to Hugging Face Dataset, loads the data\n",
+    "    datasets = load_dataset(path)\n",
+    "    return datasets\n",
+    "\n",
+    "\n",
+    "def tokenize_function(dat):\n",
+    "    tokenizer = AutoTokenizer.from_pretrained(MODEL_CHECKPOINT)\n",
```
Comment (matallanas) [inline starter_training_script.ipynb]: MODEL_CHECKPOINT is a global variable as you are using a function. Can you put a variable in the definition and assign a default value? It is better if anyone wanted to change the value

File: starter_training_script.ipynb [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+    "    return tokenizer(dat[\"text\"])\n",
+    "\n",
+    "\n",
+    "def group_texts(dat):\n",
```
Comment (matallanas) [inline starter_training_script.ipynb]: The same in here you have been using global variables inside a function scope. I will create a default value in the parameters of the function to be able to change it.

File: starter_training_script.ipynb [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+    "    result = {\n",
+    "        k: [t[i : i + BLOCK_SIZE] for i in range(0, total_length, BLOCK_SIZE)]\n",
+    "        for k, t in concatenated_examples.items()\n",
+    "    }\n",
+    "    result[\"labels\"] = result[\"input_ids\"].copy()\n",
+    "    return result\n",
+    "\n",
+    "\n",
+    "\n",
+    "def compile_model(MODEL_CHECKPOINT):\n",
```
Comment (matallanas) [inline starter_training_script.ipynb]: This function you have the MODEL_CHECKPOINT as a parameter but can you change by a a lower case attribute so you can call it inside the function?
In addition, the optimizer parameters are not available if anyone wanted to tune them. Can you make them accesible?

File: starter_training_script.ipynb [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+      "Using custom data configuration kpriyanshu256--whisper-transcripts-b6043c49b20f7241\n",
+      "Found cached dataset json (/Users/ArjunPatel/.cache/huggingface/datasets/kpriyanshu256___json/kpriyanshu256--whisper-transcripts-b6043c49b20f7241/0.0.0/e6070c77f18f01a5ad4551a8b7edfba20b8438b7cad4d94e6ad9378022ce4aab)\n",
+      "100%|██████████| 1/1 [00:00<00:00, 358.55it/s]\n"
+     ]
+    }
+   ],
+   "source": [
+    "# tokenize and preprocess dataset\n",
+    "datasets = retrieve_dataset(DATASET_PATH)\n",
+    "\n",
```
Comment (matallanas) [inline starter_training_script.ipynb]: Remove the extra lines.

File: starter_training_script.ipynb [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+      "All the layers of TFGPT2LMHeadModel were initialized from the model checkpoint at distilgpt2.\n",
+      "If your task is similar to the task the model of the checkpoint was trained on, you can already use TFGPT2LMHeadModel for predictions without further training.\n",
+      "/Users/ArjunPatel/opt/anaconda3/envs/whisper-gpt/lib/python3.10/site-packages/keras/optimizers/optimizer_v2/adam.py:114: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.\n",
+      "  super().__init__(name, **kwargs)\n",
+      "No loss specified in compile() - the model's internal loss computation will be used as the loss. Don't panic - this is a common way to train TensorFlow models in Transformers! To disable this behaviour please pass a loss argument, or explicitly pass `loss=None` if you do not want your model to compute a loss.\n"
+     ]
+    }
+   ],
+   "source": [
+    "\n",
```
Comment (matallanas) [inline starter_training_script.ipynb]: Remove also here the extra lines

File: starter_training_script.ipynb [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+    "# apply tokenization\n",
+    "tokenized_datasets = datasets.map(\n",
+    "    tokenize_function, batched=True, num_proc=4, remove_columns = [\"text\", \"id\", \"segments\"])\n",
```
Comment (matallanas) [inline starter_training_script.ipynb]: Put the parameters into one line each.

File: starter_training_script.ipynb [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+    "model = compile_model(MODEL_CHECKPOINT)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 65,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from transformers import DefaultDataCollator\n",
```
Comment (matallanas) [inline starter_training_script.ipynb]: Take this import and put it at the beggining

File: starter_training_script.ipynb [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+      "INFO:tensorflow:Assets written to: trained_model/assets\n"
+     ]
+    }
+   ],
+   "source": [
+    "model.save(\"trained_model\")"
+   ]
+  },
+  {
+   "cell_type": "code",
```
Comment (matallanas) [inline starter_training_script.ipynb]: Remove this last cell

**Comment by matallanas** [comment]
Everything seems fine but some changes can mbe made to improve the notebook.
In addition, can you:
* move the notebook to a folder called `notebooks/`?
* add some text cell explaining what you are doing to improve readability?

**Review by matallanas — CHANGES_REQUESTED** [review]

File: starter_training_script.ipynb (hunk 1 above)
Comment (arjunpatel7) [reply starter_training_script.ipynb]: Yep, done!

File: starter_training_script.ipynb (hunk 3 above)
Comment (arjunpatel7) [reply starter_training_script.ipynb]: This is a really good point, thank you! Change made, and will be available in  next commit.

File: starter_training_script.ipynb (hunk 5 above)
Comment (arjunpatel7) [reply starter_training_script.ipynb]: Yes, I can surface them! For now, I will surface the ones that were used in this call. Later, we can add additional parameters as necessary.

**Comment by arjunpatel7** [comment]
@matallanas, thank you for your detailed and helpful feedback! All noted changes have been made. Please let me know if further are necessary, thanks!

**Comment by matallanas** [comment]
Everything seem perfect now :) Thanks for applying all the changes. Now it seems good.

**Pull request merged**